generator/CPP11/include_v2.0/msgmap.hpp

```diff
@@ -0,0 +1,236 @@
+
+#pragma once
+
+#include <endian.h>
+#include <type_traits>
+
+namespace mavlink {
+
+/**
+ * Serialization helper wrapper for mavlink_message_t
+ */
+class MsgMap {
+public:
+
+	explicit MsgMap(mavlink_message_t *p) :
+		msg(p), cmsg(p), pos(0)
+	{ }
+
+	explicit MsgMap(mavlink_message_t &p) :
+		msg(&p), cmsg(&p), pos(0)
+	{ }
+
+	explicit MsgMap(const mavlink_message_t *p) :
+		msg(nullptr), cmsg(p), pos(0)
+	{ }
+
+	inline void reset()
+	{
+		pos = 0;
+	}
+
+	inline void reset(uint32_t msgid, uint8_t len)
+	{
+		assert(msg);
+
+		msg->msgid = msgid;	// necessary for finalize
+		msg->len = len;		// needed only for deserialization w/o finalize
+		pos = 0;
+	}
+
+	template<typename _T>
+	void operator<< (const _T data);
+
+	template<class _T, size_t _Size>
+	void operator<< (const std::array<_T, _Size> &data);
+
+	template<typename _T>
+	void operator>> (_T &data);
+
+	template<class _T, size_t _Size>
+	void operator>> (std::array<_T, _Size> &data);
+
+private:
+	mavlink_message_t *msg;		// for serialization
+	const mavlink_message_t *cmsg;	// for deserialization
+	size_t pos;
+
+	template<typename _T, typename _Tin>
+	inline void msg_swap_memcpy(_T &buf, _Tin data);
+
+	template<typename _T, typename _Tout>
+	inline void cmsg_memcpy_bzero_swap_set_data(_T &buf, _Tout &data);
+};
+
+} // namespace mavlink
+
+// implementation
+
+template<typename _T, typename _Tin>
+void mavlink::MsgMap::msg_swap_memcpy(_T &buf, _Tin data)
+{
+	if (std::is_floating_point<_Tin>::value)
+		buf = *static_cast<const _T *>(static_cast<const void *>(&data));
+	else
+		buf = data;
+
+	// htoleXX functions may be empty macros,
+	// switch will be optimized-out
+	switch (sizeof(_T)) {
+	case 2:
+		buf = htole16(buf);
+		break;
+
+	case 4:
+		buf = htole32(buf);
+		break;
+
+	case 8:
+		buf = htole64(buf);
+		break;
+
+	default:
+		assert(false);
+	}
+
+	memcpy(&_MAV_PAYLOAD_NON_CONST(msg)[pos], &buf, sizeof(buf));
+}
+
+template<typename _T>
+void mavlink::MsgMap::operator<< (const _T data)
+{
+	assert(msg);
+	assert(pos + sizeof(_T) <= MAVLINK_MAX_PAYLOAD_LEN);
+
+	switch (sizeof(_T)) {
+	case 1:
+		_MAV_PAYLOAD_NON_CONST(msg)[pos] = data;
+		break;
+
+	case 2:
+		uint16_t data_le16;
+		msg_swap_memcpy(data_le16, data);
+		break;
+
+	case 4:
+		uint32_t data_le32;
+		msg_swap_memcpy(data_le32, data);
+		break;
+
+	case 8:
+		uint64_t data_le64;
+		msg_swap_memcpy(data_le64, data);
+		break;
+
+	default:
+		assert(false);
+	}
+
+	//std::cout << "M< s: " << sizeof(_T) << " p: " << pos << " d: " << data << std::endl;
+
+	pos += sizeof(_T);
+}
+
+template<class _T, size_t _Size>
+void mavlink::MsgMap::operator<< (const std::array<_T, _Size> &data)
+{
+	for (auto &v : data)
```

Comment:
I don't think there is a fixed code style, but I far prefer to always have parenthesis even in single line statement. I prevent many error now and in the future !